docs/design/2022-06-06-Adding-Index-Acceleration.md

```diff
@@ -0,0 +1,99 @@
+4. TiKV will finally combine the full copy of index data that was ingested with delta index part that was updated by user DML transaction during full copy preparation period. As the index at this time is almost completed its build process, so TiKV work in this step is not much.
+
+Since Lightning writes index data firstly into local storage instead of committing backfill transactions to TiKV. This removes conflict with user transactions and there is no transaction commit time for backfill also. The main path bottleneck of adding indexes is removed. 
+
+In addition to performance enhancement, the new solution could further lower the impact that DDL caused on DML. This point will be outstanding in the future implementation plan.
+
+## Uniqueness check
+For a unique index / primary key, we should be able to detect if a duplicate key is generated and reports an error if so. In order to archive that, we could reuse the `DuplicateDetect` interface provided by the Lightning engine:
+
+```golang
+service ImportSST {
+   ...
+   // Open a write stream to generate sst files
+   rpc Write(stream WriteRequest) returns (WriteResponse) {}
+   rpc RawWrite(stream RawWriteRequest) returns (RawWriteResponse) {}
+   ...
+ 
+   // Collect duplicate data from TiKV.
+   rpc DuplicateDetect(DuplicateDetectRequest) returns (stream DuplicateDetectResponse) {}
+}
+```
+Duplicate key will be reported when two unique index records with the same key but different value are encountered. In this case, we could report the error and cancel this DDL job.
+
+### Checkpoints
+In the current implementation, the progress of creating an index is persisted in reorgCtx, it is updated when each batch of the index is backfilled. For the new backfiller in this proposal, we could mark the progress in a similar way when an SST file generated from a batch of indexes is ingested successfully.
+      
+## Temporary files
+The temporarily generated SST files are stored in local storage, the path of storage and maximum disk usage size should be configurable, but it’s not a good idea to modify them at runtime. If size of temporary files closes to allowed maximum size, the largest existing files are ingested to TiKV and cleaned.
+
+Cleanup of the temporary files should be managed for the following cases:
+- When the job is done or rollbacked, they are cleaned by the backfiller worker.
+- Each TiDB instance should be able to check the inactive SST files and clean them, in case ownership of the job execution changes because of TiDB instance crashed.
+
+## Limitations and future works
+Compared with the current one, the CPU and memory consumption of new backfiller is not controlled smoothly when doing the ‘sort and ingest’.
+
+## Compatibility
+As an optimization, this proposal does not lead to any semantic change so theoretically there will be no compatibility issue.
+
+### Enable/Disable
+1. Use tidb_ddl_enable_fast_reorg to enable and disable new solution, default value is off.
```

Comment:
IMHO it's better to reformat it in the following way:

1. Use `tidb_ddl_enable_fast_reorg` to enable and disable the new solution, the default value is off. Example:

```
TiDB> set global tidb_ddl_enable_fast_reorg = on/off;
```

Same as item 2 and 3.